### PR TITLE
discv5: add wire protocol test vectors

### PR DESCRIPTION
**Pull request by AgeManning**
## Overview

This adds a collection of test vectors to the discv5 wire protocol. Although not entirely comprehensive, it should help new implementations conform to the specification and current implementations by passing these tests.

### CODE REVIEW

File: discv5/discv5-wire-test-vectors.md [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+#### Input Parameters
+
+    id: 1
+    total: 2
+    enr-1: "enr:-HW4QBzimRxkmT18hMKaAL3IcZF1UcfTMPyi3Q1pxwZZbcZVRI8DC5infUAB_UauARLOJtYTxaagKoGmIjzQxO2qUygBgmlkgnY0iXNlY3AyNTZrMaEDymNMrg1JrLQB2KTGtv6MVbcNEVv0AHacwUAPMljNMTg"
+    enr-2: "enr:-HW4QNfxw543Ypf4HXKXdYxkyzfcxcO-6p9X986WldfVpnVTQX1xlTnWrktEWUbeTZnmgOuAY_KUhbVV1Ft98WoYUBMBgmlkgnY0iXNlY3AyNTZrMaEDDiy3QkHAxPyOgWbxp5oF1bDdlYE6dLCUUp8xfVw50jU"
+
+#### Expected Output
+
+    nodes-response-rlp: 0x04f8f80102b8f4f8f2b877f875b8401ce2991c64993d7c84c29a00bdc871917551c7d330fca2dd0d69c706596dc655448f030b98a77d4001fd46ae0112ce26d613c5a6a02a81a6223cd0c4edaa53280182696482763489736563703235366b31a103ca634cae0d49acb401d8a4c6b6fe8c55b70d115bf400769cc1400f3258cd3138b877f875b840d7f1c39e376297f81d7297758c64cb37dcc5c3beea9f57f7ce9695d7d5a67553417d719539d6ae4b445946de4d99e680eb8063f29485b555d45b7df16a1850130182696482763489736563703235366b31a1030e2cb74241c0c4fc8e8166f1a79a05d5b0dd95813a74b094529f317d5c39d235
```
Comment (zilm13) [inline discv5/discv5-wire-test-vectors.md]: I'm not getting the same result. Could we clarify it with Geth? Some nested rlp structures are not straight-forward to encode, so I'm not sure, who have the right result here.

File: discv5/discv5-wire-test-vectors.md [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+## Packet Encodings
+
+This section provides test vectors for each individual packet type. These tests
+primarily test for correct RLP encoding in the packet types.
+
+### Random Packet
+
+#### Input Parameters
+
+    tag: 0x0101010101010101010101010101010101010101010101010101010101010101
```
[the remark this answers is not in the document]
Comment (AgeManning) [reply discv5/discv5-wire-test-vectors.md]: Agree, these are some simple quick things I had access to easily. I think we need to add a lot more tests to be comprehensive.

File: discv5/discv5-wire-test-vectors.md [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+    auth-tag: 0x020202020202020202020202
+    id-nonce: 0x0303030303030303030303030303030303030303030303030303030303030303
```
[the remark this answers is not in the document]
Comment (AgeManning) [reply discv5/discv5-wire-test-vectors.md]: It's not needed here. Copy/paste error I think.

File: discv5/discv5-wire-test-vectors.md [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+
+    random-packet-rlp: 0x01010101010101010101010101010101010101010101010101010101010101018c0202020202020202020202020404040404040404040404040404040404040404040404040404040404040404040404040404040404040404
+
+### WHOAREYOU Packet
+
+#### Input Parameters
+
+    magic: 0x0101010101010101010101010101010101010101010101010101010101010101
+    token: 0x020202020202020202020202
+    id-nonce: 0x0303030303030303030303030303030303030303030303030303030303030303
```
[the remark this answers is not in the document]
Comment (AgeManning) [reply discv5/discv5-wire-test-vectors.md]: Yep, good catch.

File: discv5/discv5-wire-test-vectors.md [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+
+This section provides test vectors for individual protocol messages defined in
+the wire protocol. These tests primarily verify the RLP encoding of each
+protocol message.
+
+### Ping Request
+
+#### Input Parameters
+
+    id: 1
```
[the remark this answers is not in the document]
Comment (AgeManning) [reply discv5/discv5-wire-test-vectors.md]: Good point. In my impl I have it as a u64, which when RLP encoded gives the byte length required to fit the number. I'd also prefer hex input here and was an oversight by me here.

File: discv5/discv5-wire-test-vectors.md (hunk 1 above)
Comment (AgeManning) [reply discv5/discv5-wire-test-vectors.md]: Now that you bring this up, I realise I've not tested this one with geth. 
We should also add the case when no ENR is returned to check that encoding. 
The spec is [... , [ENR, .. ] ] . I've built an ENR list, encoded it, then add it as a single field in the outer list.

**Comment by AgeManning** [comment]
Thanks for the feedback! I've corrected some of the easier ones, and hopefully when I get some more time, I may able to add to these. 

I'll let you know when I get the chance to compare the nested ENR encodings with geth and can hopefully resolve our mismatch.

File: discv5/discv5-wire-test-vectors.md (hunk 1 above)
Comment (fjl) [reply discv5/discv5-wire-test-vectors.md]: @zilm13 your version is correct (tip: use cmd/rlpdump from go-ethereum codebase to check)

```text
04
[
  01,
  02,
  [
    [
      1ce2991c64993d7c84c29a00bdc871917551c7d330fca2dd0d69c706596dc655448f030b98a77d4001fd46ae0112ce26d613c5a6a02a81a6223cd0c4ed\
aa5328,
      01,
      "id",
      "v4",
      "secp256k1",
      03ca634cae0d49acb401d8a4c6b6fe8c55b70d115bf400769cc1400f3258cd3138,
    ],
    [
      d7f1c39e376297f81d7297758c64cb37dcc5c3beea9f57f7ce9695d7d5a67553417d719539d6ae4b445946de4d99e680eb8063f29485b555d45b7df16a\
185013,
      01,
      "id",
      "v4",
      "secp256k1",
      030e2cb74241c0c4fc8e8166f1a79a05d5b0dd95813a74b094529f317d5c39d235,
    ],
  ],
]
```

Comment (AgeManning) [reply discv5/discv5-wire-test-vectors.md]: Yep, have corrected the test vector.

File: discv5/discv5-wire-test-vectors.md [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+##### Input Parameters
+
+    secret-key: 0x7e8107fe766b6d357205280acf65c24275129ca9e44c0fd00144ca50024a1ce7
+    id-nonce: 0xe551b1c44264ab92bc0b3c9b26293e1ba4fed9128f3c3645301e8e119f179c65
+    ephemeral-pubkey: 0xb35608c01ee67edff2cffa424b219940a81cf2fb9b66068b1cf96862a17d353e22524fbdcdebc609f85cbd58ebe7a872b01e24a3829b97dd5875e8ffbc4eea81
+    enr: []
+
+##### Expected Output
+
+    auth-pt: 0xf84405b840f753ac31b017536bacd0d0238a1f849e741aef03b7ad5db1d4e64d7aa80689931f21e590edcf80ee32bb2f30707fec88fb62ea8fbcd65b9272e9a0175fea976bc0
```
Comment (zilm13) [inline discv5/discv5-wire-test-vectors.md]: As idNonceSig is incorrect, this one is too.
`0xf84405b840f753ac31b017536bacd0d0238a1f849e741aef03b7ad5db1d4e64d7aa80689931f21e590edcf80ee32bb2f30707fec88fb62ea8fbcd65b9272e9a0175fea976b80` for me

Comment (fjl) [reply discv5/discv5-wire-test-vectors.md]: Signature value depends on how signing is implemented. The output will match only if your library implements deterministic signatures, but that's not a requirement.

File: discv5/discv5-wire-test-vectors.md [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+
+The `local-secret-key` is the raw secp256k1 private key used to sign the nonce.
+
+    id_nonce: 0x02a77e3aa0c144ae7c0a3af73692b7d6e5b7a2fdc0eda16e8d5e6cb0d08e88dd04
+    ephemeral-key: 0x9961e4c2356d61bedb83052c115d311acb3a96f5777296dcf297351130266231503061ac4aaee666073d7e5bc2c80c3f5c5b500c1cb5fd0a76abbb6b675ad157
+    local-secret-key: 0xfb757dc581730490a1d7a00deea65e9b1936924caaea8f44d476014856b68736
+
+#### Expected Outputs
+
+    id-nonce-sig: 0x3b7b8ce9df3fbd9b6367c365622ccc82a2cb9d94219401e7b08e3194f9f835764a07caad38bf0f5a7a89501a8156bb053c880774502f5cd8a6190fbe374adc89
```
[the remark this answers is not in the document]
Comment (AgeManning) [reply discv5/discv5-wire-test-vectors.md]: Sorry. This was a copy-paste error. The result I was showing was for an id-nonce consisting of all one's where the nonce here is the secret-key from the previous test vector. 
The secret key is actually 33 bytes and the nonce should be 32. I've updated this test, making the nonce 32 bytes.

File: discv5/discv5-wire-test-vectors.md (hunk 6 above)
Comment (AgeManning) [reply discv5/discv5-wire-test-vectors.md]: This is not an error due to nonce-sig mismatch. In fact, we agree on the nonce sig. The difference between our results is the rlp encoding of an empty ENR. You result is encoding an empty enr as an empty string `0x80`, whereas I (and I assume geth) are encoding it as an empty list [] `0xc0`. I assume the empty list is the correct encoding.

File: discv5/discv5-wire-test-vectors.md [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+This combines the previously generated authentication header with encryption of
+the protocol message, providing the final rlp-encoded message with an
+authentication header.
+
+##### Input Parameters
+
+    tag: 0x93a7400fa0d6a694ebc24d5cf570f65d04215b6ac00757875e3f3a5f42107903
+    auth-header-rlp: 0xf8cc8c27b5af763c446acd2749fe8ea0e551b1c44264ab92bc0b3c9b26293e1ba4fed9128f3c3645301e8e119f179c658367636db840b35608c01ee67edff2cffa424b219940a81cf2fb9b66068b1cf96862a17d353e22524fbdcdebc609f85cbd58ebe7a872b01e24a3829b97dd5875e8ffbc4eea81b856570fbf23885c674867ab00320294a41732891457969a0f14d11c995668858b2ad731aa7836888020e2ccc6e0e5776d0d4bc4439161798565a4159aa8620992fb51dcb275c4f755c8b8030c82918898f1ac387f606852
+    encryption-key: 0x9f2d77db7004bf8a1a85107ac686990b
+    message-plaintext: 0x01c20101
```
[the remark this answers is not in the document]
Comment (AgeManning) [reply discv5/discv5-wire-test-vectors.md]: Yep, thanks. Added

File: discv5/discv5-wire-test-vectors.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (zilm13) [reply discv5/discv5-wire-test-vectors.md]: Checked updated vector: matches for me

File: discv5/discv5-wire-test-vectors.md (hunk 6 above)
Comment (zilm13) [reply discv5/discv5-wire-test-vectors.md]: @AgeManning nice catch, fixed. Everything matches for me since now.

**Comment by fjl** [comment]
Thank you!

**Pull request merged**